tsdb/index/labelvalues.go

```diff
@@ -0,0 +1,245 @@
+	if r.version == FormatV1 {
+		e := r.postingsV1[name]
+		if len(e) == 0 {
+			return storage.EmptyLabelValues()
+		}
+		vals := make([]string, 0, len(e))
+		for v := range e {
+			vals = append(vals, v)
+		}
+		slices.Sort(vals)
+		return &intersectLabelValuesV1{
+			e:      e,
+			values: vals,
+			cloner: NewPostingsCloner(postings),
+			b:      r.b,
+			dec:    r.dec,
+		}
+	}
+
+	e := r.postings[name]
+	if len(e) == 0 {
+		return storage.EmptyLabelValues()
+	}
+
+	d := encoding.NewDecbufAt(r.b, int(r.toc.PostingsTable), nil)
+	// Skip to start
+	d.Skip(e[0].off)
+	lastVal := e[len(e)-1].value
+
+	return &intersectLabelValues{
+		d:       &d,
+		b:       r.b,
+		dec:     r.dec,
+		lastVal: lastVal,
+		cloner:  NewPostingsCloner(postings),
+	}
+}
+
+type intersectLabelValuesV1 struct {
+	e      map[string]uint64
+	values []string
+	cloner *PostingsCloner
+	b      ByteSlice
+	dec    *Decoder
+	cur    string
+	err    error
+}
+
+func (it *intersectLabelValuesV1) Next() bool {
+	// Look for a value with intersecting postings
+	for len(it.values) > 0 {
+		val := it.values[0]
+		it.values = it.values[1:]
+
+		postingsOff := it.e[val]
+		// Read from the postings table.
+		d := encoding.NewDecbufAt(it.b, int(postingsOff), castagnoliTable)
+		_, curPostings, err := it.dec.Postings(d.Get())
+		if err != nil {
+			it.err = fmt.Errorf("decode postings: %w", err)
+			return false
+		}
+
+		postings := it.cloner.Clone()
+		if checkIntersection(curPostings, postings) {
+			it.cur = val
+			return true
+		}
+	}
+
+	return false
+}
+
+func (it *intersectLabelValuesV1) At() string {
+	return it.cur
+}
+
+func (it *intersectLabelValuesV1) Err() error {
+	return it.err
+}
+
+func (it *intersectLabelValuesV1) Warnings() annotations.Annotations {
+	return nil
+}
+
+func (it *intersectLabelValuesV1) Close() error {
+	return nil
+}
+
+type intersectLabelValues struct {
+	d         *encoding.Decbuf
+	b         ByteSlice
+	dec       *Decoder
+	cloner    *PostingsCloner
+	lastVal   string
+	skip      int
+	cur       string
+	exhausted bool
+	err       error
+}
+
+func (it *intersectLabelValues) Next() bool {
+	if it.exhausted {
+		return false
+	}
+
+	for !it.exhausted && it.d.Err() == nil {
+		if it.skip == 0 {
+			// These are always the same number of bytes,
+			// and it's faster to skip than to parse.
+			it.skip = it.d.Len()
+			// Key count
+			it.d.Uvarint()
+			// Label name
+			it.d.UvarintBytes()
+			it.skip -= it.d.Len()
+		} else {
+			it.d.Skip(it.skip)
+		}
+
+		// Label value
+		v := yoloString(it.d.UvarintBytes())
+
+		postingsOff := it.d.Uvarint64()
+		// Read from the postings table
+		d2 := encoding.NewDecbufAt(it.b, int(postingsOff), castagnoliTable)
+		_, curPostings, err := it.dec.Postings(d2.Get())
+		if err != nil {
+			it.err = fmt.Errorf("decode postings: %w", err)
+			return false
+		}
+
+		it.exhausted = v == it.lastVal
+
+		postings := it.cloner.Clone()
+		if checkIntersection(curPostings, postings) {
```

Comment:
As the corollary to the other two comments, if we rename `s/cloner/postings/` and we use `Intersect` then this could just be:
```suggestion
		if Intersect(it.postings.Clone(), postings).Next() {
```

Reply:
The reason for implementing `checkIntersection` instead of using `Intersect` is to avoid having to allocate an iterator. I tried benchmarking your change, the mean results vs before the change are as follows:

* sec/op    => +10.82%
* B/op      => +26.29%
* allocs/op => +50.28%

I don't think some simplification of the code is worth this drop in performance.